### PR TITLE
Leaner string handling

### PR DESCRIPTION
**Pull request by mara004**
This PR attempts to remove ctypesgen's bloated Python string code and leave string handling entirely to ctypes and the caller. It requires explicit encoding/decoding of strings.

I think removing this hard to maintain and wonky piece of code would improve robustness of ctypesgen bindings. It also eliminates the incompatibility with Python 3.7.6 and 3.8.1.

See https://github.com/ctypesgen/ctypesgen/issues/77#issuecomment-1320953521

-------------

I'm quite new to the ctypesgen codebase, so it is possible that the PR still contains mistakes.
It appears to work nicely for pypdfium2, however.

On the way, I left some tasks/notes in the codebase and made a few other changes that are not strictly related to the subject. I hope this is OK, as always splitting off minor improvements is too tedious.

### CODE REVIEW

**Comment by mara004** [comment]
Hmm, flake8 doesn't seem to like my code style. I'll try to improve this.

File: ctypesgen/printer_python/printer.py [hunk 1]
```diff
@@ -144,47 +148,20 @@ def print_preamble(self):
...
-            if x is None:
-                return cls()
-            else:
-                return x
-
-        p.from_param = classmethod(from_param)
-
-    return p
-
-"""
```
Comment (mara004) [inline ctypesgen/printer_python/printer.py:177]: I removed this passage because it's not clear why this should be done with separate preamble while it's not with embedded preamble. Besides, it's unclear what the mentioned bugs are, and I'm not sure if this code might cause undesired side effects.

File: tests/testsuite.py [hunk 2]
```diff
@@ -197,6 +196,12 @@ def test_one_import(self):
         m = b.struct_mystruct()
         b.bar(ctypes.byref(m))
 
+    # NOTE(geisserml) test_two_import passed before (while test_two_import_with_embedded_preamble
+    # already failed). Not sure why. Anyway, if we submit a struct of module 'b' to a function of
+    # module 'a', it's quite logical that ctypes raises a type error. Why were the tests written
+    # like this in the first place?
+
+    @unittest.expectedFailure
```
Comment (mara004) [inline tests/testsuite.py:204]: I think that this didn't fail in the fist place might have been a side effect of the `POINTER` monkeypatch (see above)?

File: ctypesgen/printer_python/printer.py (hunk 1 above)
Comment (mara004) [reply ctypesgen/printer_python/printer.py:177]: Ah, I see this has been discussed in https://github.com/ctypesgen/ctypesgen/issues/86

Comment (mara004) [inline ctypesgen/printer_python/printer.py:177]: I'll revert the changes to `--no-embed-premable` given that it's not related to strings and the way it was before is intentional, though I admittedly don't understand yet how this monkeypatch happens to fix #86.

File: ctypesgen/ctypedescs.py [hunk 3]
```diff
@@ -276,12 +276,8 @@ def __init__(self, restype, parameters, variadic, attrib=dict()):
...
             self.errcheck = CtypesPointerCast(CtypesSpecial("c_void_p"))
 
-        # Return "String" instead of "POINTER(c_char)"
-        if self.restype.py_string() == "POINTER(c_char)":
-            if "const" in self.restype.qualifiers:
-                self.restype = CtypesSpecial("c_char_p")
-            else:
-                self.restype = CtypesSpecial("String")
+        if (self.restype.py_string() == "POINTER(c_char)"):
+            self.restype = CtypesSpecial("c_char_p")
```
Comment (mara004) [inline ctypesgen/ctypedescs.py:280]: Hmm, according to the [ctypes docs](https://docs.python.org/3.7/library/ctypes.html#ctypes.c_char_p), it's not generally possible to substitute `POINTER(c_char)` with `c_char_p` - only for null-terminated strings, so I guess this clause will have to be removed?

Comment (mara004) [reply ctypesgen/ctypedescs.py]: Done.

**Comment by nilason** [comment]
Thanks for your contribution! I will give this a good round of testing as soon as I can.
By the way, did you test this with Python 3.8.1 or 3.7.6?

**Comment by mara004** [comment]
Thanks for the response.

No, I didn't actually test with one of the concerned Python versions yet, but as the problem was the declaration `String(MutableString, Union)`, which this change removes, I'm fairly confident it should work.

However, the problem is, I don't have any C development background, and I'm not sure if the backwards-incompatible approach I am suggesting here will be acceptable for the ctypesgen community at large.
While the change plays well with pypdfium2, I guess it will make ctypesgen bindings less convenient for some other use cases.

**Comment by mara004** [comment]
A point I wasn't sure about is if/when `POINTER(c_char)` can be substituted with `c_char_p`?
`c_char_p` would be much more convenient as restype because of ctypes autoconversion, but I was under the impression it couldn't be done in general since the returned data might not be null-terminated. Is that correct, or is returned `char *` data always null-terminated in practice?

File: ctypesgen/printer_python/preamble.py [hunk 4]
```diff
@@ -417,10 +62,13 @@ def __call__(self, *args):
 
 def ord_if_char(value):
```
Comment (mara004) [inline ctypesgen/printer_python/preamble.py]: I think it might make sense to remove this function, too. In my case, it's put in at places where it doesn't make sense, and all the repeated instance checks perhaps have a performance impact.
IMO, it would be much nicer to just let the user call `ord()` where it's desired.

Example (macro):
```python
# fpdf_edit.h: 15
def FPDF_ARGB(a, r, g, b):
    return (uint32_t (ord_if_char((((((uint32_t (ord_if_char(b))).value & 0xff) | (((uint32_t (ord_if_char(g))).value & 0xff) << 8)) | (((uint32_t (ord_if_char(r))).value & 0xff) << 16)) | (((uint32_t (ord_if_char(a))).value & 0xff) << 24))))).value
```

Comment (mara004) [reply ctypesgen/printer_python/preamble.py]: Done. New example macro:
```python
# fpdf_edit.h: 15
def FPDF_ARGB(a, r, g, b):
    return uint32_t(((((uint32_t(b).value & 0xff) | ((uint32_t(g).value & 0xff) << 8)) | ((uint32_t(r).value & 0xff) << 16)) | ((uint32_t(a).value & 0xff) << 24))).value
```

**Comment by nilason** [comment]
Note:
https://github.com/ctypesgen/ctypesgen/blob/12854ed91ded078b5742f858c2c1a79bc69f2fc3/pyproject.toml#L13-L14

**Comment by mara004** [comment]
Ah, another black problem. Since the recent changes, some code now fits on a single line.
I'll push a commit shortly.

File: demo/pydemolib.py [hunk 5]
```diff
@@ -6,387 +6,30 @@
 Do not modify this file.
 """
```
Comment (mara004) [inline demo/pydemolib.py:8]: Perhaps it would be best to remove this `pydemolib.py` file altogether? Always keeping it in sync is a bit tedious, and as the C binary needs to be compiled anyway, running ctypesgen too shouldn't be much of a problem.

File: ctypesgen/printer_python/preamble.py (hunk 4 above)
Comment (mara004) [reply ctypesgen/printer_python/preamble.py]: There are still some unnecessary layers of brackets I didn't manage to get rid of...

File: ctypesgen/printer_python/printer.py [hunk 6]
```diff
@@ -3,17 +3,18 @@
...
 
 THIS_DIR = os.path.dirname(__file__)
-CTYPESGEN_DIR = os.path.join(THIS_DIR, os.path.pardir)
-PREAMBLE_PATH = os.path.join(THIS_DIR, "preamble.py")
-DEFAULTHEADER_PATH = os.path.join(THIS_DIR, "defaultheader.py")
-LIBRARYLOADER_PATH = os.path.join(CTYPESGEN_DIR, "libraryloader.py")
+CTYPESGEN_DIR = join(THIS_DIR, os.path.pardir)
+PREAMBLE_PATH = join(THIS_DIR, "preamble.py")
+DEFAULTHEADER_PATH = join(THIS_DIR, "defaultheader.py")
+LIBRARYLOADER_PATH = join(CTYPESGEN_DIR, "libraryloader.py")
```
Comment (TheCodeArtist) [inline ctypesgen/printer_python/printer.py:17]: Why this approach? 
_(is it the pythonic way?
or does it help address any linter warnings about exceeding max chars in a line? 
or something else?)_

- If not,
  let's skip this change for now.

- If yes, 
  as this is existing code (unrelated to this "lean string handling" PR)
  possibly this change belongs in it own separate commit/PR.

File: ctypesgen/printer_python/printer.py [hunk 7]
```diff
@@ -425,6 +415,7 @@ def print_simple_macro(self, macro):
         # The macro translator makes heroic efforts but it occasionally fails.
         # We want to contain the failures as much as possible.
         # Hence the try statement.
+        # TODO(geisserml) Avoid try/except wrapper for simple constant definitions
```
[the remark this answers is not in the document]
Comment (TheCodeArtist) [reply ctypesgen/printer_python/printer.py:418]: It looks cumbersome to exclude all such macros individually (if there are more than a handful?)

Would it be a good idea to **first start by logging a warning**  before `pass`, whenever we fail to handle a macro here ?

File: ctypesgen/printer_python/printer.py (hunk 6 above)
Comment (mara004) [reply ctypesgen/printer_python/printer.py:17]: Path joining is needed rather often, so the `os.path` prefix is a bit tedious.
I fixed up a path joining case below  (`f"{dst}/ctypes_preamble.py"`, which isn't really elegant or windows friendly) and then did this on the way. Explaining or even splitting up all such minor points is too much work.

@TheCodeArtist Please stop comments like this that only take up my time while no one benefits.

Comment (TheCodeArtist) [reply ctypesgen/printer_python/printer.py:17]: >  which isn't really elegant or windows friendly
i mention because i have seen `os.path.join` preferred instead of having to lookup what `join()` it is.
(especially as `os.path.join()` handles cross-platform paths properly with the appropriate path-separator)

Also, it would be easier to review/merge PR if such cleanup/useful (but not relevant) changes are kept separate.

> that only take up my time while no one benefits.
You do. i do. ctypesgen does.
In this case reducing all parts not relevant to string cleanup will help shrink this PR.
(which appears to be one of the challenges preventing it from being reviewed/accepted)

Just FYI, in all my comments, please try to have it that i am only trying to help.

Anyways, please carry on with the good work.
Don't mind me too much. 😅

File: ctypesgen/printer_python/printer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mara004) [reply ctypesgen/printer_python/printer.py:418]: Just realized macro exclude already exists (`--undefine` option).

I believe the macro translator rarely fails, so no try/except and individual excludes (or manual fixup) should be fine. Otherwise, I can also limit the change to integer constants for now and else keep the try/except.

**Comment by mara004** [comment]
On comment https://github.com/ctypesgen/ctypesgen/pull/162#issuecomment-1328316567

> A point I wasn't sure about is if/when `POINTER(c_char)` could be substituted with `c_char_p`.

Revisiting the situation for return strings, I think we should not use `c_char_p` directly since ctypes restype autoconversion would try to resolve that to Python `bytes`, which may not be generally doable. But nevertheless, `POINTER(c_char)` is very inconvenient, so I think a small helper class might do good in this place.
I have in mind something that can be treated either as pointer or as bytes (via `c_char_p.value`), at the caller's choice.
Maybe also with a convenience method to resolve to a utf-8 str directly.

**Comment by mara004** [comment]
Coming back to this, it's clear enough to me that this breaking change is unlikely to be accepted.
I'm thinking about nevertheless resubmitting this as optional, and replacing the current strings classes with something lean just to maintain UTF-8 autoconversion.

**Comment by mara004** [comment]
Closing in favour of #177. However, this still contains some misc style improvements / TODOs to pick up eventually.
